File: src/javascript/algorithms/array/spiralOrderMatrix2.js
const spiralOrderMatrix2 = (A) => {
  const matrix = Array.from({ length: A }, () => Array(A).fill(0));
  let top = 0;
  let bottom = A - 1;
  let left = 0;
  let right = A - 1;

  let current = 1;

  while (current <= A * A) {
    for (let i = left; i <= right; i++) {
      matrix[top][i] = current;
      current++;
    }
    top++;

    for (let i = top; i <= bottom; i++) {
      matrix[i][right] = current;
      current++;
    }
    right--;

    for (let i = right; i >= left; i--) {
      matrix[bottom][i] = current;
      current++;
    }
    bottom--;

    for (let i = bottom; i >= top; i--) {
      matrix[i][left] = current;
      current++;
    }
    left++;
  }

  return matrix;
};

// Example usage:
const A = 4;
console.log(spiralOrderMatrix2(A));

// Time Complexity - O(A * A)
// Space Complexity - O(A * A)

File: src/java1/algorithms/array/SpiralOrderMatrix2.java
package java1.algorithms.array;

import java.util.Arrays;

public class SpiralOrderMatrix2 {

    public static int[][] spiralOrderMatrix2(int A) {
        int[][] matrix = new int[A][A];
        int top = 0;
        int bottom = A - 1;
        int left = 0;
        int right = A - 1;

        int current = 1;

        while (current <= A * A) {
            // Fill from left to right along the top row
            for (int i = left; i <= right; i++) {
                matrix[top][i] = current;
                current++;
            }
            top++;

            // Fill from top to bottom along the rightmost column
            for (int i = top; i <= bottom; i++) {
                matrix[i][right] = current;
                current++;
            }
            right--;

            // Fill from right to left along the bottom row
            for (int i = right; i >= left; i--) {
                matrix[bottom][i] = current;
                current++;
            }
            bottom--;

            // Fill from bottom to top along the leftmost column
            for (int i = bottom; i >= top; i--) {
                matrix[i][left] = current;
                current++;
            }
            left++;
        }

        return matrix;
    }

    public static void main(String[] args) {
        int A = 4;
        int[][] result = spiralOrderMatrix2(A);

        // Print the generated matrix
        for (int[] row : result) {
            System.out.println(Arrays.toString(row));
        }
    }

}

// Time Complexity - O(A * A)
// Space Complexity - O(A * A)